keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,309 @@
+
+
+
+![](./bucket-greenfield-api-relations.png)
+
+### Brownfield: Dynamic Bucket Access
+
+The expected workflow for interacting with a pre-existing storage instance is similar to [Greenfield](#). The key distinction in this scenario is that a Bucket Class will reference a single, pre-existing bucket directly (indicated by the bold dashed line).  The key values for specifying storage instances should be made in the BucketClass’s parameters.  
+
+BucketContent objects will be created for Buckets, with the connection information for the storage instance populated from the BucketClass’s .  As part of this operation, the identity specified in the Bucket’s Secret should be granted permission to access the storage instance.  This operation should be configurable through the BucketClass parameters.
+
+
+
+![](./bucket-brownfield-api-relations.png)
+
+
+
+### Brownfield: Driverless Bucket Access
+
+Clusters with a bucket controller deployed but without a provisioner to manage backend bucket operations may still support pseudo-dynamic bucket access.  Similar to the [Brownfield: Dynamic Bucket Access](#greenfield-dynamic-bucket-creation), the BucketClass will specify a single pre-existing bucket.
+
+The key difference in this case is that without a driver to communicate with an object store, operations such as adding an identity to an ACL or managing bucket polices are not possible.  As such, applications must have credentials that are permissioned to access the backing bucket.  The API will serve purely as a representation of the bucket.
+
+### Manually Managed Buckets
+
+Clusters with zero automation around the API still benefit from it when deploying Bucket aware workload-orchistrating controllers.   This case is releveant as it’s likely to be the first step towards adoption for some users.  The app will In this case, admins would define BucketContents themselves and users would create Buckets with a prefilled `bucketContentRef` specifying the BucketContent.  After that, a workload controller would be ready to inject connection and credential data from the BucketContent into the worker pods.
+
+![](/Users/jcope/Workspace/go/src/github.com/yard-turkey/enhancements/keps/sig-storage/bucket-static-api-relation.png)
+
+
+
+### Provisioner Secrets
+
+Per [Non-Goals](#non-goals), it is not within the scope of the proposal to abstract IAM operations.  Instead, provisioner and user credentials should be provided to automation by admins or users.  
+
+BucketClass.Parameter key values are used to specify Secret name and namespace.  These may be considered of different degrees of granularity.
+
+- **Per Provisioner:** the Secret is used for all provisioning operations.  These Secrets should be injected directly in provisioner containers via [common Kubernetes patterns](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/).
```

Comment:
How would this work if the `Secret` namespace/name are defined in a `BucketClass` `Parameter` field?

Reply:
Do you mean "how do we enforce a heirarchy"? This would ultimately be on admins and driver authors.  We'd have no knowledge of what secrets are mounted to a driver's container.  So, if an admin specifies a secret in the BucketClass, and the driver can accept the credentials via an interface, then it's up to the driver to decide what has priority.  

Admins would have to be aware of driver capabilities in this case.

Reply:
So, a driver should still be able to retrieve `Secret`s from the K8s API?

Let me expand on my first comment: if the `BucketClass` specifies a `Secret` to use (name/namespace), then how could this be injected in the K8s `Pod` where the provisioner is running, unless 'something else' creates this provisioner `Pod(Spec)` to include this `Secret` as a mounted volume, or an environment variable, or whatever? Basically, if the provisioner is already running and should manage multiple `BucketClass`es, how would this work?

Furthermore, exposing a `Secret` as a volume, or environment variables, or whatnot, IIRC only works when the `Pod` sits in the same `Namespace` as the `Secret`, which seems restrictive.